synapseclient/models/agent.py

```diff
@@ -0,0 +1,629 @@
+        """Converts the request to a request expected of the Synapse REST API."""
+        return {
+            "concreteType": self.concrete_type,
+            "sessionId": self.session_id,
+            "chatText": self.prompt,
+            "enableTrace": self.enable_trace,
+        }
+
+    def fill_from_dict(self, synapse_response: Dict[str, str]) -> "AgentPrompt":
+        """
+        Converts a response from the REST API into this dataclass.
+
+        Arguments:
+            synapse_response: The response from the REST API.
+
+        Returns:
+            The AgentPrompt object.
+        """
+        self.id = synapse_response.get("jobId", None)
+        self.session_id = synapse_response.get("sessionId", None)
+        self.response = synapse_response.get("responseText", None)
+        return self
+
+    async def _post_exchange_async(
+        self, *, synapse_client: Optional[Synapse] = None, **kwargs
+    ) -> None:
+        """Retrieves information about the trace of this prompt with the agent.
+
+        Arguments:
+            synapse_client: If not passed in and caching was not disabled by
+                    `Synapse.allow_client_caching(False)` this will use the last created
+                    instance from the Synapse class constructor.
+        """
+        if self.enable_trace:
+            trace_response = await get_trace(
+                prompt_id=self.id,
+                newer_than=kwargs.get("newer_than", None),
+                synapse_client=synapse_client,
+            )
+            self.trace = trace_response["page"][0]["message"]
+
+
+@dataclass
+@async_to_sync
+class AgentSession(AgentSessionSynchronousProtocol):
+    """Represents a [Synapse Agent Session](https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/agent/AgentSession.html)
+
+    Attributes:
+        id: The unique ID of the agent session.
+            Can only be used by the user that created it.
+        access_level: The access level of the agent session.
+            One of PUBLICLY_ACCESSIBLE, READ_YOUR_PRIVATE_DATA,
+            or WRITE_YOUR_PRIVATE_DATA.
+        started_on: The date the agent session was started.
+        started_by: The ID of the user who started the agent session.
+        modified_on: The date the agent session was last modified.
+        agent_registration_id: The registration ID of the agent that will
+            be used for this session.
+        etag: The etag of the agent session.
+
+    Note: It is recommended to use the `Agent` class to start and get sessions,
+    but you are free to use this class directly if you wish.
+
+    Example: Start a session and send a prompt.
+
+        from synapseclient import Synapse
+        from synapseclient.models.agent import AgentSession, AgentSessionAccessLevel
+
+        syn = Synapse()
+        syn.login()
+
+        my_session = AgentSession(agent_registration_id=my_agent_registration_id).start(
+            synapse_client=syn
+        )
```

Comment:
Since `synapse_client` is optional (And only required for advanced use cases where you might be using it in a service like Schematic), I had been leaving it out of examples for the most part.